atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/DefaultShardProgressUpdater.java

```diff
@@ -0,0 +1,151 @@
+    }
+
+    /**
+     * Returns a {@link BucketProbeResult} indicating a prefix of cells that have been swept successfully, starting
+     * from a given point in time. It is assumed that all buckets before searchStart have been swept successfully;
+     * if this is not the case, behaviour is undefined.
+     */
+    private BucketProbeResult findCompletedBuckets(ShardAndStrategy shardAndStrategy, long searchStart) {
+        for (long offset = 0; offset < MAX_BUCKETS_TO_CHECK_PER_ITERATION; offset++) {
+            long currentBucket = searchStart + offset;
+            Optional<BucketProgress> bucketProgress =
+                    bucketProgressStore.getBucketProgress(Bucket.of(shardAndStrategy, currentBucket));
+            if (bucketProgress.isPresent()) {
+                BucketProgress presentBucketProgress = bucketProgress.get();
+                TimestampRange requiredRange = recordsTable.get(shardAndStrategy, currentBucket);
+                if (presentBucketProgress.timestampProgress()
+                        != requiredRange.endExclusive() - requiredRange.startInclusive() - 1) {
+                    // Bucket still has progress to go, so we can stop here.
+                    return BucketProbeResult.builder()
+                            .endExclusive(currentBucket)
+                            .knownSweepProgress(
+                                    requiredRange.startInclusive() + presentBucketProgress.timestampProgress())
+                            .build();
+                } else {
+                    // Bucket fully processed, keep going!
+                    if (offset == MAX_BUCKETS_TO_CHECK_PER_ITERATION - 1) {
+                        // We actually finished a full set of buckets and all were completed.
+                        return BucketProbeResult.builder()
+                                .endExclusive(currentBucket + 1)
+                                .knownSweepProgress(requiredRange.endExclusive() + 1)
+                                .build();
+                    }
+                }
+            } else {
+                // No progress; we're ahead of the read pointer, so interpret as unstarted.
+                return BucketProbeResult.builder()
+                        .endExclusive(currentBucket)
+                        .knownSweepProgress(recordsTable
+                                        .get(shardAndStrategy, currentBucket)
+                                        .startInclusive()
+                                - 1L)
+                        .build();
+            }
+        }
+        throw new SafeIllegalStateException("Didn't expect to get here");
+    }
+
+    private long getStrictUpperBoundForSweptBuckets(ShardAndStrategy shardAndStrategy) {
+        Set<Bucket> startingBuckets =
+                sweepBucketPointerTable.getStartingBucketsForShards(ImmutableSet.of(shardAndStrategy));
+        if (startingBuckets.isEmpty()) {
```

Comment:
The cardinality of the startingBuckets must always match the set of shards and strategies (we return 0 in cases where the entry for starting bucket is not found), so we should just throw.

I'm tracking writing up the guarantees for each method.